src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -281,6 +281,15 @@ internal int ReadCore(Span<byte> buffer)
                     int n = _stream.Read(_buffer, 0, _buffer.Length);
                     if (n <= 0)
                     {
+                        // - Inflater didn't return any data although a non-empty output buffer was passed by the caller.
+                        // - More input is needed but there is no more input available.
+                        // - Inflation is not finished yet.
+                        // - Provided input wasn't completely empty
+                        // In such case, we are dealing with a truncated input stream.
+                        if (!buffer.IsEmpty && !_inflater.Finished() && _inflater.NonEmptyInput())
```

Comment:
If I understand correctly, the `!buffer.IsEmpty` check prevents failures on `System.IO.Compression.Tests.ZipFile_Create.CreateFromDirectory_IncludeBaseDirectory`. 
The scenario in that test is:
1. You depleted the `BaseStream`, which is used to feed the inflater. 
2. You pass an empty buffer, then the code will attempt to read from the depleted BaseStream, because the inflater thinks that it `NeedsInput()`
3. Then you check `!_inflater.Finished() && _inflater.NonEmptyInput()`, which at that point is true.

That suggests to me that what we actually want to guard here is, `!inflater.IsFinished()` AND `_inflater.NonEmptyInput()` AND `_inflater.StillContainsOutput()`. 

Isn't there a way to determine if the inflater can keep returning data? I don't think `!buffer.IsEmpty` is a reliable way of doing it.

@stephentoub @danmoseley @carlossanlop thoughts?

Reply:
> If I understand correctly, the !buffer.IsEmpty check prevents failures on System.IO.Compression.Tests.ZipFile_Create.CreateFromDirectory_IncludeBaseDirectory.

Thanks to @marcin-krystianc, this case is now also covered in the `StreamTruncation_IsDetected` test with the `Read, ReadAsync` scenarios.

> I don't think !buffer.IsEmpty is a reliable way of doing it.

Why not? I had used `_inflater.AvailableOutput > 0` before which also seemed correct.

Reply:
I've been rethinking it and I think it makes sense to not throw for `buffer.IsEmpty`, we don't pinvoke `inflate` in such case so there's no way to tell if there's still available output on the inflater.